bump httpcomponents versions

ahgittin opened the pull request:
note core and client are no longer kept in sync; latest version of each has been taken

this seems to fix strange problems with `HttpRequestSensorYamlTest` failing in maven but working in eclipse, presumably due to use of `TestHttpServer` using these components.  it does not however fix the JVM crash in that test under java 7 sadly.

aledsage commented:
This is probably ok, but we should double-check the conflicting dependencies (e.g. will it cause runtime errors in `httpasyncclient:jar:4.1` (a dependency of `cxf-rt-transports-http-hc`, which is needed by winrm4j)). There may well be other conflicts as well:
```
[INFO] |  |  +- org.apache.brooklyn:brooklyn-software-winrm:jar:0.10.0-SNAPSHOT:compile
[INFO] |  |  |  +- (org.apache.brooklyn:brooklyn-core:jar:0.10.0-SNAPSHOT:compile - omitted for duplicate)
[INFO] |  |  |  +- io.cloudsoft.windows:winrm4j:jar:0.4.0:compile
[INFO] |  |  |  |  \- io.cloudsoft.windows:winrm4j-client:jar:0.4.0:compile
[INFO] |  |  |  |     +- org.apache.cxf:cxf-rt-transports-http:jar:3.1.4:compile
[INFO] |  |  |  |     |  \- org.apache.cxf:cxf-core:jar:3.1.4:compile
[INFO] |  |  |  |     |     +- org.codehaus.woodstox:woodstox-core-asl:jar:4.4.1:compile
[INFO] |  |  |  |     |     |  \- org.codehaus.woodstox:stax2-api:jar:3.1.4:compile
[INFO] |  |  |  |     |     \- org.apache.ws.xmlschema:xmlschema-core:jar:2.2.1:compile
[INFO] |  |  |  |     +- org.apache.cxf:cxf-rt-transports-http-hc:jar:3.1.4:compile
[INFO] |  |  |  |     |  +- (org.apache.cxf:cxf-core:jar:3.1.4:compile - omitted for duplicate)
[INFO] |  |  |  |     |  +- (org.apache.cxf:cxf-rt-transports-http:jar:3.1.4:compile - omitted for duplicate)
[INFO] |  |  |  |     |  +- org.apache.httpcomponents:httpcore-nio:jar:4.4.1:compile
[INFO] |  |  |  |     |  |  \- (org.apache.httpcomponents:httpcore:jar:4.4.1:compile - omitted for conflict with 4.4.5)
[INFO] |  |  |  |     |  \- org.apache.httpcomponents:httpasyncclient:jar:4.1:compile
[INFO] |  |  |  |     |     +- (org.apache.httpcomponents:httpcore:jar:4.4.1:compile - omitted for conflict with 4.4.5)
[INFO] |  |  |  |     |     +- (org.apache.httpcomponents:httpcore-nio:jar:4.4.1:compile - omitted for duplicate)
[INFO] |  |  |  |     |     \- (org.apache.httpcomponents:httpclient:jar:4.4.1:compile - omitted for conflict with 4.5.2)
```

Also, does this work in karaf? Or might some dependent bundles have been built to depend on httpclient `[4.4,4.5)` (so fail to install, unless we include both versioned bundles in the karaf install)?

neykov commented:
Another reason to upgrade - current version doesn't use system proxy passed through `-Djava.net.useSystemProxies=true` or `-Dhttp.proxyHost`

neykov commented:
Can apply the changes on `feature.xml` as well.

neykov commented:
LGTM. Feels like the right moment to merge this.

neykov commented:
I have some failing tests locally:
```
  ApplicationResourceTest.testDeployApplicationFromBuilder:227->BrooklynRestResourceTest.waitForApplicationToBeRunning:157 expected [true] but found [false]
  UsageResourceTest.testGetApplicationUsage:161->createApp:309->waitForTask:351 task AsOhrOHp not finished expected [true] but found [false]
  UsageResourceTest.testListApplicationUsages:90->createApp:309->waitForTask:351 task R1AyyoGm not finished expected [true] but found [false]
  UsageResourceTest.testListMachinesUsageForApp:302 usages=[UsageStatistics{statistics=[UsageStatistic{status=RUNNING, id='oawydt4n04', applicationId='oawydt4n04', start='2016-12-16T19:12:57+0000', end='2016-12-16T19:12:57+0000', duration=7, metadata={}}], links={}}] expected [true] but found [false]

```

ahgittin commented:
recent changes:  fixed one compatibility issue, refer to v0.11 not 0.10, and rebased on master . (see last two commits for these tiny recent changes)

i don't seem to get the failing tests @neykov reports so maybe that was a different issue

re karaf and impact on "httpasyncclient:jar:4.1 (a dependency of cxf-rt-transports-http-hc, which is needed by winrm4j)", as noted by @aledsage, i have not tested either.  i'll try to do this but others (cc @bostko re winrm4j) may be more efficient at doing that which is fine also!

neykov commented:
I've tested this with winrm4j and it's not working. It's sending the wrong `Content-Type`, resulting in a `415` response code.

Bumping the cxf version to `3.1.9` works fine though and it brings in almost the same dependencies:
```
[INFO] +- org.apache.cxf:cxf-rt-transports-http-hc:jar:3.1.9:compile
[INFO] |  +- org.apache.httpcomponents:httpcore-nio:jar:4.4.4:compile
[INFO] |  |  \- org.apache.httpcomponents:httpcore:jar:4.4.4:compile
[INFO] |  \- org.apache.httpcomponents:httpasyncclient:jar:4.1.2:compile
[INFO] |     \- org.apache.httpcomponents:httpclient:jar:4.5.2:compile
```

Suggest upgrading `cxf` to `3.1.9` and downgrading core to `4.4.4`.

neykov commented:
See https://github.com/cloudsoft/winrm4j/pull/44 for another reason to upgrade to CXF 3.1.9.

aledsage commented on karaf/features/src/main/feature/feature.xml:
```diff
@@ -51,8 +51,8 @@
         <bundle dependency="true">mvn:org.yaml/snakeyaml/${snakeyaml.version}</bundle>
-        <bundle dependency="true">mvn:org.apache.httpcomponents/httpcore-osgi/${httpclient.version}</bundle>
-        <bundle dependency="true">mvn:org.apache.httpcomponents/httpclient-osgi/${httpclient.version}</bundle>
+        <bundle dependency="true">mvn:org.apache.httpcomponents/httpcore-osgi/${httpclient.httpcore.version}</bundle>
```
Should this not be `${httpcomponents.httpcore.version}` (and the line below `httpcomponents.httpclient.version`)?

aledsage commented:
Trying to test this (including the cxf upgrade to 3.1.9), it is failing when starting karaf with:
```
2017-01-06 15:29:40,191 ERROR  11 o.a.k.f.i.s.BootFeaturesInstaller [pool-7-thread-1] Error installing boot features
org.osgi.service.resolver.ResolutionException: Unable to resolve org.ops4j.pax.web.pax-web-jsp/4.2.4: missing requirement [org.ops4j.pax.web.pax-web-jsp/4.2.4] osgi.wiring.package; filter:="(&(osgi.wiring.package=javax.servlet.jsp.resources)(version>=2.2.0))"
        at org.apache.felix.resolver.ResolutionError.toException(ResolutionError.java:42)[org.apache.felix.framework-5.4.0.jar:]
        at org.apache.felix.resolver.ResolverImpl.resolve(ResolverImpl.java:404)[org.apache.felix.framework-5.4.0.jar:]
        at org.apache.felix.resolver.ResolverImpl.resolve(ResolverImpl.java:158)[org.apache.felix.framework-5.4.0.jar:]
        at org.apache.karaf.features.internal.region.SubsystemResolver.resolve(SubsystemResolver.java:216)[11:org.apache.karaf.features.core:4.0.4]
        at org.apache.karaf.features.internal.service.Deployer.deploy(Deployer.java:263)[11:org.apache.karaf.features.core:4.0.4]
        at org.apache.karaf.features.internal.service.Deployer.deploy(Deployer.java:259)[11:org.apache.karaf.features.core:4.0.4]
        at org.apache.karaf.features.internal.service.FeaturesServiceImpl.doProvision(FeaturesServiceImpl.java:1089)[11:org.apache.karaf.features.core:4.0.4]
        at org.apache.karaf.features.internal.service.FeaturesServiceImpl$1.call(FeaturesServiceImpl.java:985)[11:org.apache.karaf.features.core:4.0.4]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)[:1.8.0_111]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)[:1.8.0_111]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)[:1.8.0_111]
        at java.lang.Thread.run(Thread.java:745)[:1.8.0_111]
```
Talking to @m4rkmckenna, it is likely that the error shown is not the underlying cause.

I'm going to try applying these changes incrementally (starting with the cxf 3.1.9 upgrade), to see if I can get karaf to start properly!

aledsage commented:
See https://issues.apache.org/jira/browse/BROOKLYN-423 for tracking the upgrade to cxf 3.1.9, and some initial findings. I'll look into this a little longer.

ahgittin commented:
#547 is a better more holistic fix; closing here in preference to that
